src/rgw/rgw_data_sync.cc

```diff
@@ -1315,15 +1315,26 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
     reenter(this) {
       if (state->obligation) {
         // this is already syncing in another DataSyncSingleEntryCR
-        if (state->obligation->timestamp < obligation.timestamp) {
-          // cancel existing obligation and overwrite it
-          tn->log(10, SSTR("canceling existing obligation " << *state->obligation));
-          complete = std::move(*state->obligation);
-          *state->obligation = std::move(obligation);
-          state->counter++;
+        if (!state->obligation->gen || !obligation.gen || 
+            (state->obligation->gen >= obligation.gen)) {
+          // one if the generations is missing, or new obligation has older/same generation
```

Comment:
if i understand correctly, missing generation is either due to full sync (either because this is the first sync for the bucket, or that sync was disabled and then enabled) or admin initiated sync.
i think that in both cases we there might be a gen=0 already in the list of generations, and the only way to know which entry to take is based on timestamp. so, if this is an old full sync, we can probably cancel, and if this is a new full sync we can probably cancel the existing.